Define gzip encoding by default in the headers; properly decompress it

vkuznet opened the pull request:
Fixes #10451 

#### Status
In development

#### Description
Provides changes to request and accept gzip encoding from upstream server. The changes are applied to pucurl manager only which is suppose to be used across all WMCore projects.

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
#10872 , #10451 , https://github.com/dmwm/dasgoclient/issues/34

#### External dependencies / deployment changes
NO

vkuznet committed:
first draft of gzip changes

vkuznet commented:
@amaltaro , @todor-ivanov this PR represents first draft of changes for enforcing gzip encoding in HTTP requests via pycurl manager. If there are other ways we query services please let me know their location within WMCore. Meanwhile here is the logic:
- if user (code) who call pycurl_manager provides HTTP Accept-Header it will be acknowledged (i.e. not modified), otherwise
- code will add gzip Accept-Encoding request, this will go to upstream server, if server can handle this request the response will be encoding and Content-Encoding header will be set (this is the case of new DBS Go server)
- when response is received and if it has Content-Encoding gzip header it will be decompressed via `decompress` function which I provided in the code
- I added appropriate unit test for `decompress` function, but we'll need integration tests with new DBS server to see how it will work, and integration tests with other (Python based) servers which does not have gzip encoding.

I only left work in progress label and removed do not merge one since code passed unit tests and pylint. Please review and provide your comments.

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -56,11 +57,45 @@
+    except:
+        logger = logging.getLogger()
+        logger.error("HTTP headers %s" % headers)
+        logger.error("unable to decompress gzip content:\n%s" % traceback.format_exc())
```
Instead of using traceback, I sugges to use `logger.exception`. Maybe you want to propagate the original error message as well, so something like:
```
except Exception as exc:
    msg = "Error details: {}".format(str(exc)) etc etc etc
```

amaltaro reviewed (changes requested):
Valentin, I left a few comments along the code.

Don't we also need to implement this gzip for the `getdata` function? If not, then is there any reason not to make the `decompress` function to actually be a method of RequestHandler class?

vkuznet commented:
@amaltaro thanks for comments, I resolved all of the issues. To answer your question, the `RequestHandler` and `getdata` both use decompress to handle the gzip'ed content. We don' need dedicated function since existing code should work transparently with any HTTP requests, i.e. with and without gzip encoding. The encoding response is provided by the server if two conditions are met, the client ask for it or server can handle it. If only one condition exists, e.g. client ask for gziped content, and server does not able to process it, the server will provide content in the usual way. That's why both `RequestHandler` and `getdata` should work transparently.

todor-ivanov commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -47,20 +47,54 @@
+        msg += "we were unable to decompress gzip content:\n%"
+        msg += str(exc)
+        logger.exception(msg)
+        return body
```
This here I'd say is a good strategy to return the body (same for the `return body` above) in case the server has set the header inappropriately and the contents are actually not zipped, but then wouldn't it be worth checking the returned value bellow and if this was the exit point from the `decompress` method, then we also retry the regular: 

```
data = curl.bbuf.getvalue().decode('utf-8')
```

I am talking about lines 511-513 from the current PR. And I can see the type of behavior I am talking about is implemented at lines 393,394 inside the `multirequest` method.

todor-ivanov reviewed (commented):
Hi @vkuznet  thank you for this PR. This is indeed quite needed.

 I do see it is quite helpful to have the behavior to be set through the request header, and I do consider this would be the correct way. But sometimes the header my not reflect fact that the data could be zipped. One example I can think of is, if it is notified only through the server's API that the data is compressed, but not in the header. (e.g. a part of the `uri` is containing something like `&format=gzip` or whatever parameter the developers have decided to use on the server side). 

Of course `pycurl_manager` has no way to know that, unless we do not say that explicitly. So in this case wouldn't it be appropriate to add additional parameter so that we can provide that information at call time in the cases we already know what type of data to expect.

vkuznet commented:
@todor-ivanov , regarding your suggestion to use `&format=gzip`. No, please let's not invent the wheel and make custom behavior of handling HTTP requests. We should and must follow HTTP standard. The standard provides two HTTP headers:
- Content-Encoding which specifies how content is passed from client to the server, the server here should address content encoding appropriately.
- Accept-Encoding specified what client accepts from the server, in this case the server can decide which encoding to use based on client's request

todor-ivanov reviewed (approved):
Thanks @vkuznet  for answering my questions from the previous review. 
All looks good according to me.

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -185,6 +219,9 @@ def set_opts(self, curl, url, params, headers,
                                 type(headers["Accept-Encoding"]))
+        else:
+            # add gzip encoding by default
+            curl.setopt(pycurl.ENCODING, 'gzip')
```
Do you see any use case and/or scenario where hard-coding this `gzip` accept-encoding could be a problem for us?

vkuznet replied:
no, since it is up to a server to accept it or not. If accepted and returned back as gzip we have code in place to process it. Otherwise, the change will have no effect on our code.

amaltaro reviewed (changes requested):
Thank you for providing these changes, Valentin. I made a few suggestions and questions along the code.

Just for completeness, HTTP calls that were inhering from the Services class should be already setting the `gzip` option at the header, e.g.:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Requests.py#L174

vkuznet commented:
Alan, the code you pointing to https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Services/Requests.py#L174 is not the only place where pycurl_manager can be called. For instance, I don't know if CRAB or DBS Client call it through your Requests.py codebase. Therefore, to make changes transparent to WMCore Requests, i.e. in other words when someone will use pycurl_manager directly, we need to setup defaults within pycurl_manager.

amaltaro commented:
test this please

amaltaro commented on test/python/WMCore_t/Services_t/pycurl_manager_t.py:
```diff
@@ -25,6 +27,21 @@ def setUp(self):
+        body = "bla"
+        headers = {}
+        data = decompress(body, headers)
+        self.assertTrue(data, body)
```
Valentin, this has been marked as resolved, but this assert still needs to be fixed.

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -47,20 +47,54 @@
+        return body
+
+    try:
+        zbuf = io.BytesIO()
```
Valentin, you resolved this but didn't answer my previous question. I worry of overdoing things here in this very base class and adding a performance penalty to all HTTP requests.

In this line:
https://github.com/dmwm/WMCore/blob/81b12bc25b0ed4de536cd5d6d83970c4e010291b/src/python/WMCore/Services/pycurl_manager.py#L327

you already get the content of the (bytes) buffer as a string. Then you write that string again to an in-memory buffer object. 

Maybe we should pass the buffer over to method and return a `body.getvalue()` if it's not a gzip content?

I didn't look carefully into this gzip library as well, but is there any reason that we cannot do something like `gzip.decompress(body)`?

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -47,20 +47,52 @@
+        zbuf.seek(0)
+        with gzip.GzipFile(mode='rb', fileobj=zbuf) as zfile:
+            data = zfile.read()
+            return data
```
Line over-indentend (should be out of the context manager)

amaltaro commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -465,12 +506,11 @@ def getdata(urls, ckey, cert, headers=None, options=None, num_conn=50, cookie=No
-                    data = curl.bbuf.getvalue().decode('utf-8')
+                hdrs = curl.hbuf.getvalue().decode('utf-8')
+                if 'content-encoding: gzip' in str(hdrs).lower():
+                    data = decompress(curl.bbuf.getvalue(), ResponseHeader(hdrs).getHeader())
```
Reviving this discussion, given that `decompress` already has a check on the content type, I would suggest to remove this header check here and simply call that `decompress` function.

amaltaro reviewed (changes requested):
Valentin, I reopened some discussions because I think it's important to understand and reach an optimal implementation for such base module (the whole WMCore relies on this).

In addition to the comments made along the code, I also see an inconsistent usage of data encoding, i.e., a mix of (same for body buff - bbuf):
```
hbuf.getvalue()
and
hbuf.getvalue().decode('utf-8')
```

Last but not least, could you please clean this unused import:
```
 W0611, line 55: Unused import sys 
```
as reported by pylint (even though the change has not been made by you). Since it's one line change, please squash it with the logical changes commit).

amaltaro commented:
FYI @belforte even though it's meant to be a transparent change.

vkuznet commented:
Alan, thanks for suggestions, I'll answer your question here in one block:
- regarding `sys` module, I removed it
- regarding usage of `gzip.decompress`, I switched to it after test passed (good suggestion)
- regarding header check, I fixed it

Now, regarding your suggestion to use buffer instead of string (which we read from it). I think it is doable, but I want to do it after the above changes will pass and your review. I looked at the code and it seems we can pass the bbuf into decompress, but it will require to change `def parse_body` method as well to pass into it the buffer since it now has decompress call. This change may work or may not since I don't know who else is using `parse_body` method. It may be used outside of this class, and therefore I suggest to make bbuf change separately.

belforte commented:
thanks for keeping me in the loop. Would it make sense to look into adding gzip also on WMCore/REST server side as well ? We have a API in CRAB which often tries to get back O(10MB) of json and runs into FE 5min timeout. I wonder if this would be simpler for us than changing things at deeper level and make it need/retrieve less, or break in pieces etc. Maybe (in due time) @mapellidario can be interested ? Other parts of the system may also benefit, before/without we rewrite them in go.
Or am I simply lost ! Sorry in that case !

vkuznet commented:
@amaltaro , @todor-ivanov now tests are passed again and I'm requesting your review again. As I wrote above, the Alan's suggestion about using buffer instead of string in decompress may work but since it affect another public method of the class I rather prefer to do it separately (we can do it after your review or even in separate PR). I think I answered all Alan's question now and your feedback is really welcome.

todor-ivanov reviewed (approved):
looks good to me

amaltaro reviewed (changes requested):
Please find further comments along the code.

Note that unittest changes should have NOT been added to the same commit providing the logical/functional changes. It's annoying to separate it now, but please keep it in mind for future contributions.

vkuznet commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -61,6 +61,31 @@
+        return gzip.decompress(body)
+    except Exception as exc:
+        logger = logging.getLogger()
+        msg = "while processing decompress function with headers\n:%s" % headers
```
done

vkuznet commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -350,7 +380,9 @@ def multirequest(self, url, parray, headers=None,
             dummyNumq, response, dummyErr = multi.info_read()
             for dummyCobj in response:
-                data = json.loads(bbuf.getvalue())
+                headers = self.parse_header(hbuf.getvalue())
```
multirequest is provided if you want to send in parallel mutlple requests to the upstream server. But you're correct about decode, this code was developed in py2 era and we gradually fix the decode issues.

vkuznet commented on src/python/WMCore/Services/pycurl_manager.py:
```diff
@@ -465,12 +497,8 @@ def getdata(urls, ckey, cert, headers=None, options=None, num_conn=50, cookie=No
-                else:
-                    hdrs = curl.hbuf.getvalue()
-                    data = curl.bbuf.getvalue()
+                hdrs = curl.hbuf.getvalue().decode('utf-8')
```
Alan, please be clear, do you want to replace ALL occurrences where we call `buf.getvalue.decode('utf-8')` or this particular case. If former, it will bring much more changes (I don't mind but that was not the intention of this PR). If later it will not be consistent with the rest of the code.

vkuznet commented:
@amaltaro change to `decodeBytesToUnicode` seems fine. Please review again and hopefully we can merge it.

amaltaro reviewed (approved):
It looks good to me, Valentin. Could you please squash these 2 commits? Thanks

vkuznet commented:
squash is done!

amaltaro commented:
And we should have updated the commit message and the PR title! I am updating at least the PR title...
